### PR TITLE
Remove all `useEffect` usages

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

The primary usage of effects is to interact with an external system. We don't have this. That's why this PR replaces all
`useEffect` usages with better suited hooks or removes them alltogether. See https://react.dev/learn/you-might-not-need-an-effect

## Test Plan

* Made changes
* Reload the editor
* Verified that changes are persisted
* Changed the theme to dark mode and reloaded the editor. Theme was persisted

I wasn't able to test sharing. The wrangler server goes up to 100% CPU usage whenever I hit share with or without my changes...

### CODE REVIEW

File: playground/.gitignore
Comment (charliermarsh) [inline playground/.gitignore:132]: Extra newline here.

File: playground/src/main.tsx
Comment (charliermarsh) [inline playground/src/main.tsx:23]: Doesn't this mean that nothing on the page will render until after the data fetch?

File: playground/src/main.tsx
Comment (dhruvmanila) [inline playground/src/main.tsx:9]:
```suggestion
import init, { Workspace } from "./pkg/ruff_wasm";
```

File: playground/src/main.tsx
Comment (charliermarsh) [reply playground/src/main.tsx:23]: But, we don't even show the editor until the source is loaded, right?

File: playground/src/Editor/Editor.tsx
Comment (MichaReiser) [inline playground/src/Editor/Editor.tsx:103]: @charliermarsh I think we showed an empty editor and the `initAsync` that loads the wasm and restores the python code raced against the user making any changes :laughing:

File: playground/src/main.tsx
Comment (MichaReiser) [reply playground/src/main.tsx:23]: We did show the editor. See https://github.com/astral-sh/ruff/pull/12659/files#r1704097956

Isn't that changing this is your main concern?

Comment (charliermarsh) [reply playground/src/main.tsx:23]: I thought we showed the chrome (header) but not the editable panes in the editor, just based on reading the code where we have that `source ?` ternary in the editor body.

**Comment by charliermarsh** [comment]
I (or someone) should probably test sharing before we merge this.

File: playground/src/main.tsx
Comment (MichaReiser) [reply playground/src/main.tsx:23]: ah yeah, that's correct.

**Review by charliermarsh — APPROVED** [review]
I do think we should test sharing before merging. I can try later today.

File: playground/src/main.tsx
Comment (charliermarsh) [reply playground/src/main.tsx:23]: I defer to you, I don't want to get in your way.

Comment (MichaReiser) [reply playground/src/main.tsx]: I extracted a `Chrome` component. It would have been nice if [`React.use` ](https://react.dev/reference/react/use) weren't experimental :(

**Comment by MichaReiser** [comment]
Okay, I had to accept or decline the telemetry option, then force kill all wrangler process to get it working. But sharing works

**Pull request merged**
